### PR TITLE
Add output response parser

### PR DESCRIPTION
**Pull request by jlalmes**
## Added output validation to router queries & mutations

I have a goal to try and auto-generate OpenAPI docs for a TRPC router (cc. https://github.com/trpc/trpc/issues/755). The feature in this PR will enable us to generate JSON schemas for request and response payloads. This output parser is **opt-in only** and should not introduce any breaking changes.

### API
```
const appRouter = new trpc.router<Context>()
  .query({
    input: z.string(),
    output: z.object({
      input: z.string(),
    }),
    resolve: ({ input }) => {
      return { input };
    }
  });
```

- Added a new optional `output` property.
- When `output` is provided, the `resolve` function will expect a return type inferred from the `output` parser.
- The response data will be validated against the `output` parser, unless `NODE_ENV` is `production`. 
- If the response data fails validation a `TRPCError INTERNAL_SERVER_ERROR` error will be thrown.

### Todo
- Get some feedback (cc. @KATT)
- Write some comprehensive output parser tests.
- Update documentation.

### CODE REVIEW

**Review by KATT — COMMENTED** [review]
Not able to look it through until tomorrow at the earliest but it is great! I would love for optional output validation

**Comment by KATT** [comment]
I think if there is an output validation, it should always run, we can add the optionality as a feature later!

**Comment by KATT** [comment]
Also setup github sponsors so I can send you 💷💷!

**Comment by sachinraja** [comment]
> I have a goal to try and auto-generate OpenAPI docs for a TRPC router

Let me know if there's any way I can help! I have some experience in this area with [`zod-to-ts`](https://github.com/sachinraja/zod-to-ts) and [tRPC Playground](https://github.com/sachinraja/zod-to-ts).

File: packages/server/test/inference.test.ts [hunk 1]
```diff
@@ -16,6 +17,25 @@ test('infer input & output', async () => {
...
+      async resolve({ input }) {
+        return { input };
+      },
+    })
+    .query('preferOutput', {
+      input: z.string(),
+      output: z.object({
+        input: z.string(),
+      }),
+      // @ts-ignore - ensuring types from "output" are expected
```
Comment (KATT) [inline packages/server/test/inference.test.ts:35]:
```suggestion
      // @ts-expect-error - ensuring types from "output" are expected
```

File: packages/server/test/inference.test.ts [hunk 2]
```diff
@@ -16,6 +17,25 @@ test('infer input & output', async () => {
       async resolve({ input }) {
         return { input };
       },
+    })
+    .query('withOutput', {
+      input: z.string(),
+      output: z.object({
+        input: z.string(),
+      }),
```
Comment (KATT) [inline packages/server/test/inference.test.ts:25]: Maybe tests to do with this should be in it's own file? Like `output-parser.test.ts`

File: packages/server/src/internals/procedure.ts [hunk 3]
```diff
@@ -97,22 +86,26 @@ function getParseFn<TParsedInput>(
...
 
-  constructor(opts: ProcedureOptions<TContext, TParsedInput, TOutput>) {
+  constructor(opts: ProcedureOptions<TContext, TInput, TOutput>) {
     this.middlewares = opts.middlewares;
     this.resolver = opts.resolver;
     this.inputParser = opts.inputParser;
-    this.parse = getParseFn(this.inputParser);
+    this.parseInputFn = getParseFn(this.inputParser);
+    this.outputParser = opts.outputParser;
+    this.parseOutputFn = this.outputParser && getParseFn(this.outputParser);
```
Comment (KATT) [inline packages/server/src/internals/procedure.ts:103]: This could maybe be simplified for usage it just has a passthrough fn

```suggestion
    this.parseOutputFn = this.outputParser ?? (val) => val;
```

File: packages/server/src/internals/procedure.ts [hunk 4]
```diff
@@ -121,6 +114,20 @@ export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
     }
   }
 
+  private async parseOutput(rawOutput: unknown): Promise<TOutput> {
+    try {
+      if (process.env.NODE_ENV === 'production' || !this.parseOutputFn) {
+        return rawOutput as TOutput;
+      }
+      return await this.parseOutputFn(rawOutput);
```
Comment (KATT) [inline packages/server/src/internals/procedure.ts:122]: Would be nice if it was:

```suggestion
  return await this.parseOutputFn(rawOutput);
```

File: packages/server/src/internals/procedure.ts [hunk 5]
```diff
@@ -121,6 +114,20 @@ export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
...
+    } catch (cause) {
+      throw new TRPCError({
+        code: 'INTERNAL_SERVER_ERROR',
+        cause,
+      });
```
Comment (KATT) [inline packages/server/src/internals/procedure.ts:127]: Definitely, it's an internal server error if this is wrong. Unsure if we should do more to guide the developer.

Maybe something like this?:

```suggestion
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        cause,
        message: 'Output validation failed',
      });
```

**Review by KATT — COMMENTED** [review]
This all looks really sound to me. GOOD JOB!! 🙏 🙏 

Some suggestions:

- Make the output parser stuff into its own test file
- Don't allow bad outputs regardless of the environment
- Some smaller comments inline

**Comment by KATT** [comment]
You might've accidentally removed the #1182 feature BTW ☺

**Comment by jlalmes** [comment]
> #1182

Rolled it back - will make sure that any transforms in the output zod schemas are also handled!

**Comment by KATT** [comment]
> > #1182
> 
> Rolled it back - will make sure that any transforms in the output zod schemas are also handled!

Seems like an edge case to have transformers in the output but why not!

**Review by sachinraja — CHANGES_REQUESTED** [review]

File: packages/server/test/inference.test.ts (hunk 1 above)
Comment (jlalmes) [reply packages/server/test/inference.test.ts]: Done 👍

File: packages/server/src/internals/procedure.ts (hunk 3 above)
Comment (jlalmes) [reply packages/server/src/internals/procedure.ts]: outdated

File: packages/server/src/internals/procedure.ts (hunk 4 above)
Comment (jlalmes) [reply packages/server/src/internals/procedure.ts]: outdated

File: packages/server/src/internals/procedure.ts (hunk 5 above)
Comment (jlalmes) [reply packages/server/src/internals/procedure.ts]: done 👍

File: packages/server/test/inference.test.ts (hunk 2 above)
Comment (jlalmes) [reply packages/server/test/inference.test.ts:18]: I have kept these high-level type inference related tests in this file. However, I have also added `outputParser.test.ts` with a more comprehensive set of tests relating to this PR, hope this is ok. 👍

File: packages/server/src/internals/procedure.ts [hunk 6]
```diff
@@ -100,29 +96,49 @@ function getParseFn<TParsedInput>(
 export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
   private middlewares: Readonly<Array<MiddlewareFunction<any, any>>>;
   private resolver: ProcedureResolver<TContext, TParsedInput, TOutput>;
-  private readonly inputParser: ProcedureInputParser<TParsedInput>;
-  private parse: ParseFn<TParsedInput>;
+  public readonly inputParser: ProcedureParser<TParsedInput>;
+  private parseInputFn: ParseFn<TParsedInput>;
+  public readonly outputParser: ProcedureParser<TOutput> | undefined;
+  private parseOutputFn: ParseFn<TOutput>;
```
[the remark this answers is not in the document]
Comment (jlalmes) [reply packages/server/src/internals/procedure.ts]: I have changed `inputParser` and `outputParser` back to `private` for now as this is out of scope for this PR. I imagine we will change them to `public` to generate openapi documentation in a future PR though 👍

File: packages/server/src/internals/procedure.ts [hunk 7]
```diff
@@ -233,19 +256,14 @@ export type CreateProcedureOptions<
...
-    >
+  | CreateProcedureWithTransformInput<TContext, TInput, TParsedInput, TOutput>
   | CreateProcedureWithoutInput<TContext, TOutput>;
 
 export function createProcedure<TContext, TInput, TParsedInput, TOutput>(
   opts: CreateProcedureOptions<TContext, TInput, TParsedInput, TOutput>,
 ): Procedure<unknown, TContext, TInput, TParsedInput, TOutput> {
   const inputParser =
-    'input' in opts
+    'input' in opts && opts.input
```
[the remark this answers is not in the document]
Comment (jlalmes) [reply packages/server/src/internals/procedure.ts]: done 👍

**Comment by jlalmes** [comment]
## Update

Hi @KATT & @sachinraja - I have just updated this PR & resolved all outstanding comments, please review again and leave any feedback!

### Changes to this PR
* Ensure transformers are handled correctly in zod output schemas (https://github.com/trpc/trpc/pull/1699#issuecomment-1079649264)
* Removed `NODE_ENV === "production"` output validation bypass
* Added env var `TRPC_SKIP_OUTPUT_VALIDATION === "true"` output validation bypass
* Added `"Output validation failed"` error message on validation failure
* Reverted `outputParser` & `inputParser` back to `private` class members of Procedure (outside scope of this PR)
* Added a comprehensive tests for this PR to `outputParser.test.ts` file
* Refactoring

### Outstanding tasks
* API documentation
* Add examples

.

**Comment by jlalmes** [comment]
Just a thought - Currently `output` is omitted from `subscription()` method. 

Do you think we should allow `output` in `subscription()` and ensure that we are validating the `emit.data(output)` output payload?

**Review by KATT — COMMENTED** [review]
Played around with this and it works great as far as I can see.

Last thing - could you add a section in the docs? `/www` folder

**Review by KATT — COMMENTED** [review]
Will merge this tomorrow and do a release.

I removed the `TRPC_SKIP_OUTPUT_VALIDATION` because I don't want magic runtime variables - this should either be explicitly implemented per procedure or on the app router similarily to transformers/error formatters.

**Review by KATT — APPROVED** [review]
